### ISSUE TITLE
Change rule `no-unused-vars` to warn about unused function arguments

### ISSUE THREAD

**Issue by szkrd** [standard/standard#419]
Is it intentional that no-unused-vars are set only for variables, but not for arguments? I can live with it of course, just curious.

`"no-unused-vars": 2` vs. `"no-unused-vars": [2, { "args": "all" }]`

**Comment by LinusU** [comment]
It usually clashes with the ugly way that express defines handles error handlers...

I brought it up here https://github.com/feross/standard/issues/216#issuecomment-127569835 and here sindresorhus/xo#6

**Comment by szkrd** [comment]
Ah, now I see, thank you!

**Comment by LinusU** [comment]
It might be a good idea to open up discussion about this again, and to run thru the tests and see how many repos would be affected by changing this...

**Comment by dcousens** [comment]
@LinusU I use express.  I feel this pain.

**Comment by LinusU** [comment]
Yeah, me too. But it's literally the only place I have this problem though, which makes it a bit annoying since it could catch other bugs...

**Comment by szkrd** [comment]
We are mostly a koa shop :) Though adding it manually to eslintrc feels kinda dirty.

**Comment by dcousens** [comment]
@LinusU the other places I had this issue were eventually lobbied enough to remove it,  but I still feel its pain from time to time.

**Comment by feross** [comment]
I just opened an issue on `express` about deprecating their four-argument middleware in favor of an explicit `app.error` method: https://github.com/expressjs/express/issues/2896

**Comment by feross** [comment]
Ran `{ "args": "after-used" }` through the test suite:

```
1..427
# tests 427
# pass  322
# fail  105
```

**Comment by akivajgordon** [comment]
For me, a consequence of not being informed about unused params can cause tests to fail. As an example, I started with an asynchronous test:

```js
it('should foo', (done) => {
  somethingAsync().then(() => {
    expect(...)
    done()
  })
})
```

But then, I refactored my test and it no longer needed to be async:

```js
it('should foo', (done) => {
  doSomething()
  expect(...)
})
```

Oops! Now my test fails because I never called `done()`. Had the linter warned me about `done` being unused, I would have removed it and not been running an asynchronous test. In other words, the mere presence of the `done` argument (or lack thereof) actually changes the way the test behaves, so it would be nice if I was warned about it.

**Comment by yoshuawuyts** [comment]
I believe we used to lint for this, but then stopped doing that. Not sure
why, I missed the conversation on that one

**Comment by dcousens** [comment]
Indeed,  I am bitten by this often.
The few times it still occurs in `express`, I use `_` to indicate an unused variable.

**Comment by dcousens** [comment]
What about

"no-unused-vars": [2, { "vars": "all", "args": "all", "argsIgnorePattern": "^_$" }],

**Comment by feross** [comment]
@akivajgordon Mocha should not count the number of arguments to the function to determine whether it's async or sync. Fortunately, this kind of fanciness is going out of fashion: it's planned to be removed from express and was already removed from other libraries that do it, like superagent. It should have explicit methods for sync and async tests, IMO.

@dcousens Any idea what the ecosystem impact of that rule change would be?

@yoshuawuyts This was removed well over a year ago, I believe.

**Comment by LinusU** [comment]
It won't be a problem in the case of mocha though, since you always call the `done` callback somehow, if your test is asynchronous...

**Comment by feross** [comment]
Moving discussion to here: https://github.com/standard/eslint-config-standard/pull/136

**Comment by feross** [comment]
Once express ships https://github.com/expressjs/express/issues/2896 we can enable this rule. In the meantime, we're blocked.

**Comment by feross** [comment]
Hmm, honestly, I'm not so sure that we should discourage the inclusion of parameters even if they're unused, especially for anonymous functions used as callbacks. For example:

```js
test('no callback (array)', function (t) {
  t.plan(2)

  var tasks = [
    function (cb) {
      t.pass('cb 1')
    },
    function (cb) {
      t.pass('cb 2')
    }
  ]

  parallel(tasks)
})
```

Removing the `cb` parameter here just loses information.

On the other hand, if it's a named function, and not a callback, then I think it's useful. For example:

```js
function checkIEEE754 (buf, value, offset, ext, max, min) {
  if (offset + ext > buf.length) throw new RangeError('Index out of range')
  if (offset < 0) throw new RangeError('Index out of range')
}
```

It's confusing to have a `max` and `min` parameter that is not used.

So maybe we need an option for distinguishing between the cases?

Also, there's MASSIVE ecosystem breakage:

```
1..203
# tests 203
# pass  149
# fail  54
```

**Comment by LinusU** [comment]
```js
test('no callback (array)', function (t) {
  t.plan(2)

  var tasks = [
    function (cb) {
      t.pass('cb 1')
    },
    function (cb) {
      t.pass('cb 2')
    }
  ]

  parallel(tasks)
})
```

Seems like not using `cb` here was unintended though? If the task functions are supposed to call that function when they are done 🤔, and I'm not really sure what information is being lost by removing them 🤔 

Would be interesting to look at all the breakage to see some examples, will try and get some time for that!

**Comment by dcousens** [comment]
I understand the ecosystem impact is the blocker,  but,  `_` signals the difference between "forgotten to use" and "intentionally omitted"

**Comment by voxpelli** [comment]
What TypeScript does when one has set [`noUnusedParameters`](https://www.typescriptlang.org/tsconfig#noUnusedParameters) there is exactly what @dcousens suggest here, adding a `"argsIgnorePattern": "^_"`, but then also add an `"args": "all"` to actually ensure that one hasn't forgotten to use any other arguments either
